### PR TITLE
Remove reliance on calculated width and prevent workspace from displaying outside its container

### PR DESCRIPTION
**Pull request by jmkulwik**
Relates to this PR in the code-dot-org repo: https://github.com/code-dot-org/code-dot-org/pull/29849

**OLD:**
workspaceWithoutDropletChanges

**NEW:**
workspaceWithDropletChanges

**CS50 workspace after changes**
cs50withChanges

### CODE REVIEW

File: src/controller.coffee [hunk 1]
```diff
@@ -3651,6 +3658,7 @@ Editor::performMeltAnimation = (fadeTime = 500, translateTime = 1000, cb = ->) -
       #  @paletteWrapper.style.left = '-9999px'
 
       @dropletElement.style.top = '-9999px'
+      @dropletElement.style.bottom = '9999px'
```
Comment (cpirich) [inline src/controller.coffee]: should you also specify `right` here for symmetry?

File: src/controller.coffee [hunk 2]
```diff
@@ -3703,13 +3711,14 @@ Editor::performFreezeAnimation = (fadeTime = 500, translateTime = 500, cb = ->)-
...
       @showScrollbars false
-      @dropletElement.style.width = @wrapperElement.clientWidth + 'px'
+      @dropletElement.style.right = "0px"
 
       @redrawMain noText: true
 
       @currentlyAnimating_suppressRedraw = true
 
       @aceElement.style.top = "-9999px"
+      @aceElement.style.bottom = "9999px"
```
Comment (cpirich) [inline src/controller.coffee]: same as above, should you specify `right` as well?

File: src/controller.coffee [hunk 3]
```diff
@@ -3861,13 +3872,15 @@ Editor::enablePalette = (enabled) ->
         #@paletteWrapper.style.left = '-9999px'
 
         @currentlyAnimating = false
-
-        @redrawMain()
+        @resize()
+        @resizeNubby()
+        @paletteWrapper.style.visibility = "hidden"
```
Comment (cpirich) [inline src/controller.coffee]: I'm not totally sure I follow this particular change.

File: src/controller.coffee [hunk 4]
```diff
@@ -3948,9 +3961,8 @@ hook 'populate', 2, ->
     @session.viewports.palette.x = @paletteScroller.scrollLeft
 
 Editor::resizeMainScroller = ->
-  @mainScroller.style.width = "#{@dropletElement.clientWidth}px"
-  @mainScroller.style.height = "#{@dropletElement.clientHeight}px"
```
Comment (cpirich) [inline src/controller.coffee:3952]: was this line never needed? or is just not needed now that you are specifying `bottom` and `top` in other places?

**Review by cpirich — COMMENTED** [review]
I really like the direction of these changes (it is definitely a better, more CSS friendly way to handle sizing).

I'm a little concerned about the continued reliance on moving things "offscreen" with the large 9999px or -9999px offsets. (but that's not a new problem - just seems to appear in more places now)

And I'm a little concerned at how many different places needed to have tiny changes applied, so it isn't obvious if you've missed some. In the future, we could probably make this cleaner.

File: src/controller.coffee (hunk 4 above)
Comment (jmkulwik) [reply src/controller.coffee:3952]: top, bottom, left, and right were already set in droplet.css. This line may have been needed at some point, but isn't anymore.

File: src/controller.coffee (hunk 3 above)
Comment (jmkulwik) [reply src/controller.coffee]: (Assuming you are talking about the "visibility" change)

Good point. This was left over from an earlier attempt. Thanks for catching it. I'll remove this and the other "visibility" line.

**Review by cpirich — APPROVED** [review]
Looks good! I'm hopeful that this all works, but expect that you'll need a fair amount of manual testing (if you haven't already).

**Pull request merged**